macid.py

```diff
@@ -0,0 +1,289 @@
+#Licensed to the Apache Software Foundation (ASF) under one or more contributor license 
+#agreements; and to You under the Apache License, Version 2.0.
+
+import numpy as np
+from pgmpy.models import BayesianModel
+from pgmpy.factors.discrete import TabularCPD
+from pgmpy.factors.continuous import ContinuousFactor
+from pgmpy.factors.discrete import TabularCPD
+import logging
+import typing
+from typing import List, Tuple, Dict
+import itertools
+from pgmpy.inference import BeliefPropagation
+import functools
+import networkx as nx
+from cpd import NullCPD
+import matplotlib.pyplot as plt
+import operator
+from collections import defaultdict
+
+
+class MACID(BayesianModel):
+    def __init__(self, ebunch:List[Tuple[str, str]]=None, node_types:Dict=None, decision_cardinalities:Dict=None, utility_values:Dict=None ):
+        super(MACID, self).__init__(ebunch=ebunch)
+        self.node_types = node_types
+        self.decision_cardinalities = decision_cardinalities
+        self.utility_values = utility_values
+        self.utility_nodes = dict((i, node_types[i]['U']) for i in node_types if i != 'C')     # this gives a dictionary matching each agent with their decision and utility nodes
```

Comment:
In "examples2.py", there are a load of example MACIDs. As it's currently setup, here's an example:

{1: {'D': ['PT', 'BP'], 'U': ['E', 'V']}, 2: {'D': ['TDoc'], 'U': ['Tree', 'Cost']}, 'C': ['TS', 'TDead']},
So you have player 1's decision nodes are 'PT and 'BP'; player 1's utility nodes are 'E' and 'V'; player 2's decision node is 'TDoc'; player 2's utility nodes are 'Tree' and 'Cost'; and the chance nodes are 'TS' and 'TDead'.

Reply:
> I think it would be more natural and consistent with the existing codebase to have which decisionmaker owns each decision or utility node be a part of the NullCPD or UtilityCPD.

It doesn't seem to me to be more effort to describe node types as I have done above - ie as a dictionary?

I also thinks this makes it instantly clear what's going on when you're looking at an example.

Reply:
On 1), Can this line be `{i:node_types[i]['U'] for i in node_types if i != 'C'}`?

Reply:
Ha yes - thanks! Of course